### PR TITLE
Add support for Tensorflow SparseTensors: optimizers.

### PR DESCRIPTION
**Pull request by hertschuh**
All optimizers now handle sparse gradients in the form of `tf.IndexedSlices`. Where applicable, this results in partial updates of variables using `scatter_add`/`scatter_sub`.

This was done by performing the following changes in optimizers code:
- replacing all negative operations (e.g. `+`) with the equivalent Keras op (e.g. `ops.add`). Note that an exception was made for scalars.
- replacing `variable.assign/assign_add/assign_sub` with `self.assign/assign_add/assign_sub`.

### CODE REVIEW

**Review by fchollet — APPROVED** [review]
LGTM -- great work!
